src/libraries/Microsoft.PowerFx.Core/IR/Visitors/DependencyVisitor.cs

```diff
@@ -0,0 +1,350 @@
+        {
+            public bool WriteState { get; set; }
+
+            public TableType TableType { get; set; }
+
+            public DependencyContext()
+            {
+            }
+        }
+
+        // Translate relationship names to actual field references.
+        public string Translate(string tableLogicalName, string fieldLogicalName)
+        {
+            return fieldLogicalName;
+        }
+
+        // if fieldLogicalName, then we're taking a dependency on entire record.
+        private void AddField(Dictionary<string, HashSet<string>> list, string tableLogicalName, string fieldLogicalName)
+        {
+            if (tableLogicalName == null)
+            {
+                return;
+            }
+
+            if (!list.TryGetValue(tableLogicalName, out var fieldReads))
+            {
+                fieldReads = new HashSet<string>();
+                list[tableLogicalName] = fieldReads;
+            }
+            
+            if (fieldLogicalName != null)
+            {
+                var name = Translate(tableLogicalName, fieldLogicalName);
+                fieldReads.Add(fieldLogicalName);
+            }
+        }
+
+        public void AddFieldRead(string tableLogicalName, string fieldLogicalName)
+        {
+            if (Info.FieldReads == null)
+            {
+                Info.FieldReads = new Dictionary<string, HashSet<string>>();
+            }
+
+            AddField(Info.FieldReads, tableLogicalName, fieldLogicalName);
+        }
+
+        public void AddFieldWrite(string tableLogicalName, string fieldLogicalName)
+        {
+            if (Info.FieldWrites == null)
+            {
+                Info.FieldWrites = new Dictionary<string, HashSet<string>>();
+            }
+
+            AddField(Info.FieldWrites, tableLogicalName, fieldLogicalName);
+        }
+
+        public void AddField(DependencyContext context, string tableLogicalName, string fieldLogicalName)
+        {
+            if (context.WriteState)
+            {
+                AddFieldWrite(tableLogicalName, fieldLogicalName);
+            }
+            else
+            {
+                AddFieldRead(tableLogicalName, fieldLogicalName);
+            }
+        }
+    }
+
+    /// <summary>
+    /// Capture Dataverse field-level reads and writes within a formula.
+    /// </summary>
+    public class DependencyInfo
+    {
+        /// <summary>
+        /// A dictionary of field logical names on related records, indexed by the related entity logical name.
+        /// </summary>
+        /// <example>
+        /// On account, the formula "Name & 'Primary Contact'.'Full Name'" would return
+        ///    "contact" => { "fullname" }
+        /// The formula "Name & 'Primary Contact'.'Full Name' & Sum(Contacts, 'Number Of Childeren')" would return
+        ///    "contact" => { "fullname", "numberofchildren" }.
+        /// </example>
+        public Dictionary<string, HashSet<string>> FieldReads { get; set; }
```

Comment:
>string [](http://example.com/codeflow?start=26&length=6)

I think we will ultimately want to move away from strings.and to something more structured.